File: src/main/java/net/irisshaders/lilybot/commands/services/GitHub.java
package net.irisshaders.lilybot.commands.services;

import com.jagrosh.jdautilities.command.SlashCommand;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.events.interaction.commands.SlashCommandEvent;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import net.irisshaders.lilybot.LilyBot;
import net.irisshaders.lilybot.utils.Memory;
import org.kohsuke.github.*;

import java.awt.*;
import java.io.IOException;
import java.text.DecimalFormat;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

public class GitHub extends SlashCommand {

    public GitHub() {
        this.name = "github";
        this.help = "Gets some information from GitHub.";
        this.defaultEnabled = true;
        this.guildOnly = false;
        this.children = new SlashCommand[]{
          new GitHubIssueSubCommand(),
          new GitHubRepoSubCommand(),
          new GitHubUserSubCommand()
        };
    }

    @Override
    protected void execute(SlashCommandEvent event) {
        // ignored because all commands are sub commands
    }

    @SuppressWarnings("ConstantConditions")
    public static class GitHubIssueSubCommand extends SlashCommand {

        public GitHubIssueSubCommand() {
            this.name = "issue";
            this.help = "Gets some information about an issue from a repository on GitHub.";
            this.defaultEnabled = true;
            this.guildOnly = false;
            List<OptionData> optionData = new ArrayList<>();
            optionData.add(new OptionData(OptionType.STRING, "repo", "The repo to look up. Format: User/Repo or Org/Repo.").setRequired(true));
            optionData.add(new OptionData(OptionType.STRING, "issue", "The issue. Provide a number or a search term.").setRequired(true));
            this.options = optionData;
        }

        @Override
        protected void execute(SlashCommandEvent event) {

            GHIssue issue;

            User user = event.getUser();
            String repo = event.getOption("repo").getAsString();
            String term = event.getOption("issue").getAsString();

            try {
                issue = parseMessage(repo, term);
            } catch (Exception e) {
                MessageEmbed noMatchesEmbed = new EmbedBuilder()
                        .setDescription(
                                "Invalid issue number or an invalid repository was provided." +
                                " Please ensure the issue exists and the repository is public."
                        )
                        .setColor(Color.RED)
                        .setFooter("Requested by " + user.getAsTag(), user.getEffectiveAvatarUrl())
                        .setTimestamp(Instant.now())
                        .build();
                event.replyEmbeds(noMatchesEmbed).mentionRepliedUser(false).setEphemeral(true).queue();
                return;
            }
            if (issue == null) return;

            try {
                event.replyEmbeds(issueBuilder(issue, user)).mentionRepliedUser(false).setEphemeral(false).queue();
            } catch (Exception e) {
                e.printStackTrace();
            }

        }

        public GHIssue parseMessage(String repo, String term) throws Exception {

            GHIssue issue;
            int issueNum;

            try {
                issueNum = Integer.parseInt(term);
                issue = LilyBot.INSTANCE.gitHub.getRepository(repo).getIssue(issueNum);
            } catch (NumberFormatException e) {
                PagedIterator<GHIssue> iterator = LilyBot.INSTANCE.gitHub.searchIssues()
                        .q(term + " repo:" + repo)
                        .order(GHDirection.DESC)
                        .list()
                        ._iterator(1);
                if (iterator.hasNext()) {
                    issue = iterator.next();
                } else {
                    throw new Exception("Could not find specified issue in repo `" + repo + "`!");
                }
                int num = issue.getNumber();
                issue = LilyBot.INSTANCE.gitHub.getRepository(repo).getIssue(num);
                return issue;
            }
            return issue;

        }

        public static MessageEmbed issueBuilder(GHIssue issue, User user) {

            EmbedBuilder e = new EmbedBuilder();
            e.setTitle(issue.getTitle(), String.valueOf(issue.getHtmlUrl()));

            if (issue.getBody() != null) {
                if (issue.getBody().length() > 400) {
                    e.setDescription(issue.getBody().substring(0, 399) + "...");
                } else {
                    e.setDescription(issue.getBody());
                }
            }

            boolean open = issue.getState() == GHIssueState.OPEN;
            boolean merged = false;
            boolean draft = false;

            if (issue.isPullRequest()) {

                e.setAuthor("Information for Pull Request #" + issue.getNumber() + " in " + issue.getRepository().getFullName());

                try {
                    GHPullRequest pull = issue.getRepository().getPullRequest(issue.getNumber());
                    merged = pull.isMerged();
                    draft = pull.isDraft();
                } catch (IOException ioException) {
                    ioException.printStackTrace();
                    e.setTitle("Error!");
                    e.setDescription("Error occurred initializing Pull request. How did this happen?");
                    return e.build();
                }
            } else {
                e.setAuthor("Information for Issue #" + issue.getNumber() + " in " + issue.getRepository().getFullName());
            }

            if (merged) {
                e.setColor(Color.decode("#6f42c1"));
                e.addField("Status:", "Merged", false);
            } else if (!open) {
                e.setColor(Color.decode("#cb2431"));
                e.addField("Status:", "Closed", false);
            } else if (draft) {
                e.setColor(Color.decode("#ffffff"));
                e.addField("Status:", "Draft", false);
            } else {
                e.setColor(Color.decode("#2cbe4e"));
                e.addField("Status:", "Open", false);
            }

            try {
                GHUser author = issue.getUser();
                if (author.getName() != null) e.addField("Author:", "[" + author.getLogin() + " (" + author.getName() + ")" + "](https://github.com/" + author.getLogin() + ")", false);
                else e.addField("Author:", "[" + author.getLogin() + "](https://github.com/" + author.getLogin() + ")", false);
            } catch (IOException ioException) {
                e.addField("Author:", "Unknown Author", false);
            }

            try {
                e.setFooter("Opened");
                e.setTimestamp(issue.getCreatedAt().toInstant());
                List<CharSequence> labels = new ArrayList<>();

                for (GHLabel label : issue.getLabels()) {
                    labels.add(label.getName());
                }

                if (labels.size() > 0) e.addField("Labels:", String.join(", ", labels), false);
                List<CharSequence> assignees = new ArrayList<>();
                for (GHUser assignee : issue.getAssignees()) {
                    assignees.add(assignee.getLogin());
                }
                if (assignees.size() > 0) e.addField("Assignees:", String.join(", ", assignees), false);
            } catch (IOException exception) {
                exception.printStackTrace();
            }

            e.setFooter("Requested by " + user.getAsTag(), user.getEffectiveAvatarUrl());
            e.setTimestamp(Instant.now());

            return e.build();
        }

    }

    @SuppressWarnings("ConstantConditions")
    public static class GitHubRepoSubCommand extends SlashCommand {

        public GitHubRepoSubCommand() {
            this.name = "repo";
            this.help = "Gets some information about a repository on GitHub.";
            this.defaultEnabled = true;
            this.guildOnly = false;
            List<OptionData> optionData = new ArrayList<>();
            optionData.add(new OptionData(OptionType.STRING, "repo", "The repo to look up. Format: User/Repo or Org/Repo.").setRequired(true));
            this.options = optionData;
        }

        @Override
        protected void execute(SlashCommandEvent event) {

            User user = event.getUser();
            String repoName = event.getOption("repo").getAsString();

            if (!repoName.contains("/")) {
                MessageEmbed incorrectSyntaxEmbed = new EmbedBuilder()
                        .setDescription("""
                            Make sure your input is formatted like this:
                            Format: User/Repo or Org/Repo
                            For example: `IrisShaders/Iris`
                            """
                        )
                        .setColor(Color.RED)
                        .setFooter("Requested by " + user.getAsTag(), user.getEffectiveAvatarUrl())
                        .setTimestamp(Instant.now())
                        .build();
                event.replyEmbeds(incorrectSyntaxEmbed).mentionRepliedUser(false).setEphemeral(true).queue();
                return;
            }

            GHRepository repo;

            try {
                repo = LilyBot.INSTANCE.gitHub.getRepository(repoName);
            } catch (IOException e) {
                MessageEmbed noMatchesEmbed = new EmbedBuilder()
                        .setDescription("Invalid repository name. Make sure this repository exists!")
                        .setColor(Color.RED)
                        .setFooter("Requested by " + user.getAsTag(), user.getEffectiveAvatarUrl())
                        .setTimestamp(Instant.now())
                        .build();
                event.replyEmbeds(noMatchesEmbed).mentionRepliedUser(false).setEphemeral(true).queue();
                return;
            }

            event.replyEmbeds(getRepoData(repo)).mentionRepliedUser(false).setEphemeral(false).queue();

        }

        public static MessageEmbed getRepoData(GHRepository repo) {

            EmbedBuilder e = new EmbedBuilder();

            try {
                e.setTitle("GitHub repository Info for " + repo.getFullName(), "https://github.com/" + repo.getFullName());
                e.setDescription(repo.getDescription());
                if (repo.getLicense() != null) e.addField("License:", repo.getLicense().getName(), false);
                e.addField("Open Issues/PRs:", String.valueOf(repo.getOpenIssueCount()), false);
                e.addField("Stars:", String.valueOf(repo.getStargazersCount()), false);
                e.addField("Size:", bytesToFriendly(repo.getSize()), false);
            } catch (IOException ioException) {
                ioException.printStackTrace();
            }

            return e.build();

        }

        private static String bytesToFriendly(long bytes) {

            int k = 1024;
            String[] measure = new String[]{"B", "KB", "MB", "GB", "TB"};
            double i;

            if (bytes == 0) i = 0; else i = Math.floor(Math.log(bytes) / Math.log(k));

            DecimalFormat df = new DecimalFormat("#.##");

            return df.format(bytes / Math.pow(k, i)) + " " + measure[(int) i + 1];

        }

    }

    @SuppressWarnings("ConstantConditions")
    public static class GitHubUserSubCommand extends SlashCommand {

        public GitHubUserSubCommand() {
            this.name = "user";
            this.help = "Gets some information about a user on GitHub.";
            List<OptionData> optionData = new ArrayList<>();
            optionData.add(new OptionData(OptionType.STRING, "username", "The user to look up.").setRequired(true));
            this.options = optionData;
        }

        @Override
        protected void execute(SlashCommandEvent event) {

            GHUser ghUser = null;

            User user = event.getUser();
            String username = event.getOption("username").getAsString();

            try {
                ghUser = LilyBot.INSTANCE.gitHub.getUser(username);
            } catch (IOException e) {
                MessageEmbed noMatchesEmbed = new EmbedBuilder()
                        .setDescription("Invalid username. Make sure this user exists.")
                        .setColor(Color.RED)
                        .setFooter("Requested by " + user.getAsTag(), user.getEffectiveAvatarUrl())
                        .setTimestamp(Instant.now())
                        .build();
                event.replyEmbeds(noMatchesEmbed).mentionRepliedUser(false).setEphemeral(true).queue();
            }

            event.replyEmbeds(getUser(ghUser, user)).mentionRepliedUser(false).setEphemeral(false).queue();

        }

        public static MessageEmbed getUser(GHUser ghUser, User user) {

            EmbedBuilder e = new EmbedBuilder();

            try {
                e.setThumbnail(ghUser.getAvatarUrl());
                e.addField("Repositories:", String.valueOf(ghUser.getPublicRepoCount()), false);
                e.setFooter("Requested by " + user.getAsTag(), user.getEffectiveAvatarUrl());
                e.setTimestamp(Instant.now());

                boolean isOrg = ghUser.getType().equals("Organization");

                if (!isOrg) {
                    e.setTitle("GitHub profile for " + ghUser.getLogin(), "https://github.com/" + ghUser.getLogin());
                    e.setDescription(ghUser.getBio());
                    e.addField("Followers:", String.valueOf(ghUser.getFollowersCount()), false);
                    e.addField("Following:", String.valueOf(ghUser.getFollowingCount()), false);
                    if (ghUser.getCompany() != null) e.addField("Company:", ghUser.getCompany(), false);
                    if (!ghUser.getBlog().equals("")) e.addField("Website:", ghUser.getBlog(), false);
                    if (ghUser.getTwitterUsername() != null) e.addField("Twitter:",
                            "[@" + ghUser.getTwitterUsername() + "](https://twitter.com/" + ghUser.getTwitterUsername() + ")", false);
                } else {
                    GHOrganization org = LilyBot.INSTANCE.gitHub.getOrganization(ghUser.getLogin());
                    e.setTitle("GitHub profile for " + ghUser.getLogin(), "https://github.com/" + ghUser.getLogin());
                    e.addField("Public Members:", String.valueOf(org.listMembers().toArray().length), false);
                }

            } catch (IOException ioException) {
                ioException.printStackTrace();
            }

            return e.build();

        }

    }

}

File: src/main/kotlin/net/irisshaders/lilybot/commands/Github.kt
package net.irisshaders.lilybot.commands

import com.kotlindiscord.kord.extensions.DISCORD_RED
import com.kotlindiscord.kord.extensions.commands.Arguments
import com.kotlindiscord.kord.extensions.commands.application.slash.publicSubCommand
import com.kotlindiscord.kord.extensions.commands.converters.impl.int
import com.kotlindiscord.kord.extensions.commands.converters.impl.string
import com.kotlindiscord.kord.extensions.extensions.Extension
import com.kotlindiscord.kord.extensions.extensions.publicSlashCommand
import com.kotlindiscord.kord.extensions.sentry.BreadcrumbType
import com.kotlindiscord.kord.extensions.types.respond
import com.kotlindiscord.kord.extensions.types.respondPublic
import dev.kord.rest.builder.message.create.embed
import io.ktor.utils.io.errors.IOException
import kotlinx.datetime.Clock
import net.irisshaders.lilybot.github
import net.irisshaders.lilybot.utils.ResponseHelper

import org.kohsuke.github.GHIssue
import org.kohsuke.github.PagedIterator
import org.kohsuke.github.GHDirection
import org.kohsuke.github.GHIssueState
import org.kohsuke.github.GHPullRequest
import org.kohsuke.github.GHUser
import org.kohsuke.github.GHRepository
import org.kohsuke.github.GHLabel
import org.kohsuke.github.GHOrganization
import org.kohsuke.github.GHFileNotFoundException

import java.text.DecimalFormat
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.pow


class Github : Extension() {
    override val name = "github"

    override suspend fun setup() {
        /**
         * GitHub Issue Command
         * @author NoComment1105
         * Java version author: chalkyjeans
         */
        publicSlashCommand {
            name = "github"
            description = "The parent command for all /github commands"

            publicSubCommand(::IssueArgs) {
                name = "issue"
                description = "Look up an issue on a specific repository"

                action {
                    var issue: GHIssue?

                    try {
                        issue = github?.getRepository(arguments.repository)?.getIssue(arguments.issue)
                        if (sentry.adapter.enabled) {
                            sentry.breadcrumb(BreadcrumbType.Info) {
                                category = "command.github.issue.getIssue"
                                message = "Found issue"
                            }
                        }
                    } catch (e: NumberFormatException) {
                        val iterator: PagedIterator<GHIssue>? = github?.searchIssues()
                            ?.q("${arguments.issue} repo: ${arguments.repository}")
                            ?.order(GHDirection.DESC)
                            ?.list()
                            ?._iterator(1)

                        if (iterator!!.hasNext()) {
                            issue = iterator.next()
                        } else {
                            if (sentry.adapter.enabled) {
                                sentry.breadcrumb(BreadcrumbType.Error) {
                                    category = "command.github.issue.getIssue"
                                    message = "Unable to find issue"
                                }
                            }
                            throw Exception("Could not find specified issue in repo `${arguments.repository}`!")
                        }

                        val num: Int = issue.number
                        issue = github?.getRepository(arguments.repository)?.getIssue(num)
                    }
                    respondPublic {
                        embed {

                            val open: Boolean = issue?.state == GHIssueState.OPEN
                            var merged = false
                            var draft = false

                            if (issue!!.isPullRequest) {
                                title = issue.title
                                url = issue.htmlUrl.toString()
                                description = "**Information for Pull request #${issue.number} in ${issue.repository.fullName}**"

                                try {
                                    val pull: GHPullRequest = issue.repository.getPullRequest(issue.number)
                                    merged = pull.isMerged
                                    draft = pull.isDraft
                                } catch (ioException: IOException) {
                                    if (sentry.adapter.enabled) {
                                        sentry.breadcrumb(BreadcrumbType.Error) {
                                            category = "command.github.issue.CheckPRStatus"
                                            message = "Error initializing PR wtf"
                                        }
                                    }
                                    ioException.printStackTrace()
                                    title = "Error!"
                                    description = "Error occurred initializing Pull Request. How did this happen?"
                                    color = DISCORD_RED
                                    return@action
                                }
                            } else {
                                title = issue.title
                                url = issue.htmlUrl.toString()
                                description = "**Information for issue #${issue.number} in ${issue.repository.fullName}**"
                            }

                            field {

                                if (issue.body != null) {
                                    value = if (issue.body.length > 400) {
                                        issue.body.substring(0, 399) + "..."
                                    } else {
                                        issue.body
                                    }
                                }
                            }

                            if (merged) {
                                color = dev.kord.common.Color(111, 66, 193)
                                field {
                                    name = "Status:"
                                    value = "Merged"
                                    inline = false
                                }
                            } else if (!open) {
                                color = dev.kord.common.Color(203, 36, 49)
                                field {
                                    name = "Status:"
                                    value = "Closed"
                                    inline = false
                                }
                            } else if (draft) {
                                color = dev.kord.common.Color(255, 255, 255)
                                field {
                                    name = "Status:"
                                    value = "Draft"
                                    inline = false
                                }
                            } else {
                                color = dev.kord.common.Color(44, 185, 78)
                                field {
                                    name = "Status:"
                                    value = "Open"
                                    inline = false
                                }
                            }

                            try {
                                val author: GHUser = issue.user
                                if (author.name != null) {
                                    field {
                                        name = "Author:"
                                        value =
                                            "[" + author.login + " (" + author.name + ")](https://github.com/" + author.login + ")"
                                        inline = false
                                    }
                                } else {
                                    field {
                                        name = "Author:"
                                        value = "[" + author.login + "](https://github.com/" + author.login + ")"
                                        inline = false
                                    }
                                }
                            } catch (ioException: IOException) {
                                field {
                                    name = "Author:"
                                    value = "Unknown Author"
                                    inline = false
                                }
                            }

                            try {
                                field {
                                    name = "Opened on:"
                                    value = "${issue.createdAt}"
                                    inline = false
                                }

                                val labels = mutableListOf<CharSequence>()

                                for (label: GHLabel in issue.labels) {
                                    labels.add(label.name)
                                }

                                if (labels.size > 0) {
                                    field {
                                        name = "Labels:"
                                        value = labels.joinToString(", ")
                                        inline = false
                                    }
                                }
                            } catch (ioException: IOException) {
                                ioException.printStackTrace()
                            }
                            footer {
                                text = "Requested by: " + user.asUser().tag
                                icon = user.asUser().avatar?.url
                            }
                            timestamp = Clock.System.now()
                        }
                    }
                }
            }

            publicSubCommand(::RepoArgs) {
                name = "repo"
                description = "Search GitHub for a specific repository"

                action {

                    if (!arguments.repository.contains("/")) {
                        if (sentry.adapter.enabled) {
                            sentry.breadcrumb(BreadcrumbType.Error) {
                                category = "command.github.repository.InputCheck"
                                message = "Input missing /"
                            }
                        }
                        respond {
                            content = "ohno"
                            ResponseHelper.failureEmbed(event.interaction.getChannel(),"Make sure your input is formatted like this:", "Format: `User/Repo` or `Org/Repo` \n For example: `IrisShaders/Iris`")
                        }
                        return@action
                    }

                    var repo: GHRepository?

                    try {
                        repo = github!!.getRepository(arguments.repository)
                        if (sentry.adapter.enabled) {
                            sentry.breadcrumb(BreadcrumbType.Info) {
                                category = "command.github.repository.getRepository"
                                message = "Repository found"
                            }
                        }
                    } catch (exception: Exception) {
                        if (sentry.adapter.enabled) {
                            sentry.breadcrumb(BreadcrumbType.Error) {
                                category = "command.github.repository.getRepository"
                                message = "Repository not found"
                            }
                        }
                        respond {
                            content = "ohno"
                            ResponseHelper.failureEmbed(event.interaction.getChannel(), "Invalid repository name. Make sure this repository exists!", null)
                        }
                        repo = null
                        return@action
                    }

                    respond {
                        embed {
                            try {
                                title = "GitHub Repository Info for " + repo?.fullName
                                url = repo?.htmlUrl.toString()
                                description = repo?.description

                                if (repo!!.license != null) {
                                    field {
                                        name = "License:"
                                        value = repo.license.name
                                        inline = false
                                    }
                                }

                                field {
                                    name = "Open Issues and PRs:"
                                    value = repo.openIssueCount.toString()
                                    inline = false
                                }
                                field {
                                    name = "Forks:"
                                    value = repo.forksCount.toString()
                                    inline = false
                                }
                                field {
                                    name = "Stars:"
                                    value = repo.stargazersCount.toString()
                                    inline = false
                                }
                                field {
                                    name = "Size:"
                                    value = bytesToFriendly(repo.size)
                                    inline = false
                                }
                            } catch (ioException: IOException) {
                                ioException.printStackTrace()
                            }
                        }
                    }
                }
            }
            publicSubCommand(::UserArgs) {
                name = "user"
                description = "Search github for a User/Organisation"

                action {
                    val ghUser: GHUser?

                    try {
                        ghUser = github!!.getUser(arguments.username)
                        if (sentry.adapter.enabled) {
                            sentry.breadcrumb(BreadcrumbType.Info) {
                                category = "command.github.user.getUser"
                                message = "User found"
                            }
                        }
                    } catch (exception: GHFileNotFoundException) {
                        if (sentry.adapter.enabled) {
                            sentry.breadcrumb(BreadcrumbType.Error) {
                                category = "command.github.user.getUser"
                                message = "Unable to find user"
                            }
                        }
                        respond {
                            content = "ohno"
                            ResponseHelper.failureEmbed(event.interaction.getChannel(),"Invalid Username. Make sure this user exists", null)
                        }
                        return@action
                    }

                    try {
                        val isOrg: Boolean = ghUser?.type.equals("Organization")

                        if (!isOrg) {
                            if (sentry.adapter.enabled) {
                                sentry.breadcrumb(BreadcrumbType.Info) {
                                    category = "command.github.user.isOrg"
                                    message = "User is not Organisation"
                                    data["isNotOrg"] = ghUser?.login
                                }
                            }
                            respond {
                                embed {
                                    title = "GitHub profile for " + ghUser?.login
                                    url = "https://github.com/" + ghUser?.login
                                    description = ghUser?.bio

                                    field {
                                        name = "Followers:"
                                        value = ghUser?.followersCount.toString()
                                        inline = false
                                    }

                                    field {
                                        name = "Following:"
                                        value = ghUser?.followingCount.toString()
                                        inline = false
                                    }

                                    if (ghUser!!.company != null) {
                                        field {
                                            name = "Company"
                                            value = ghUser.company
                                            inline = false
                                        }
                                    }

                                    if (!ghUser.blog.equals("")) {
                                        field {
                                            name = "Website:"
                                            value = ghUser.blog
                                            inline = false
                                        }
                                    }

                                    if (ghUser.twitterUsername != null) {
                                        field {
                                            name = "Twitter:"
                                            value =
                                                "[@" + ghUser.twitterUsername + "](https://twitter.com/" + ghUser.twitterUsername + ")"
                                            inline = false
                                        }
                                    }
                                    footer {
                                        text = ghUser.login
                                        icon = ghUser.avatarUrl
                                    }
                                    timestamp = Clock.System.now()
                                }
                            }
                        } else {
                            if (sentry.adapter.enabled) {
                                sentry.breadcrumb(BreadcrumbType.Info) {
                                    category = "command.github.user.isOrg"
                                    message = "User is Organisation"
                                    data["isOrg"] = ghUser?.login
                                }
                            }
                            val org: GHOrganization? = github?.getOrganization(ghUser?.login)

                            respond {
                                embed {
                                    title = "GitHub profile for " + ghUser?.login
                                    url = "https://github.com/" + ghUser?.login

                                    field {
                                        name = "Public Members:"
                                        value = org!!.listMembers().toArray().size.toString()
                                        inline = false
                                    }
                                    field {
                                        name = "Repositories:"
                                        value = ghUser?.publicRepoCount.toString()
                                        inline = false
                                    }

                                    footer {
                                        text = ghUser?.login.toString()
                                        icon = ghUser?.avatarUrl
                                    }

                                    timestamp = Clock.System.now()
                                }
                            }
                        }

                    } catch (ioException: IOException) {
                        ioException.printStackTrace()
                    }
                }
            }
        }
    }

    private fun bytesToFriendly(bytes: Int): String {
        val k = 1024
        val measure = arrayOf("B", "KB", "MB", "GB", "TB")

        val i: Double = if (bytes == 0) {
            0.0
        } else {
            floor(ln(bytes.toDouble()) / ln(k.toDouble()))
        }

        val df = DecimalFormat("#.##")

        return df.format(bytes / k.toDouble().pow(i)) + " " + measure[(i + 1).toInt()]
    }

    inner class IssueArgs : Arguments() {
        val repository by string("repository", "The GitHub repository you would like to search")
        val issue by int("issue-number", "The issue number you would like to search for")
    }

    inner class RepoArgs : Arguments() {
        val repository by string("repository", "The github repository you would like to search for")
    }

    inner class UserArgs : Arguments() {
        val username by string("username", "The name of the User/Organisation you wish to search for")
    }
}